File: src/y05.clj
(println
 (do

   (def error (fn [n] (throw (Throwable. "SHOULD NOT CALL"))))

   (def fp
        (fn [recursive]
          (fn [n] (if (zero? n) 1 (* n (recursive (- n 1)))))))

   (def f0 (fp error))
   (def f1 (fp f0))
   (def f2 (fp f1))

   (f1 1)

   ))

File: src/y05.js
// Fixpoints
// Higher Order Functions
// Functional Refactorings
// (1) Tennentt Correspondence Principle
// (2) Introduce binding
// (3) Rebind
// (4) Function inline

error = function(n) { throw Error("SHOULD NOT BE CALLED") }

fact_improver = function(partial) {
  return function(n) {
    return n == 0 ? 1 : n * partial(n-1)
  }
}

y = function(g) {
  return function(f) { return function(n) { return g(f(f))(n) } } (
         function(f) { return function(n) { return g(f(f))(n) } } )
}

fact = y(fact_improver)

result = fact(5)

console.log(result)
